File: src/com/colony/ArduinoSerialInterface.java
package com.colony;


import com.colony.sensor.SensorEvent;
import com.colony.sensor.Sonar;
import com.colony.serial.SerialPortIO;
import com.colony.serial.SerialPortListener;
import com.colony.servo.Devices;
import io.reactivex.Observer;

/**
 * get serial data from arduino connected to usb port and post feedback to main program
 */

public class ArduinoSerialInterface {





    public void startListening(Observer<SensorEvent> observer) {

        SerialPortIO serialPortIO = new SerialPortIO("ttyACM0", new SerialPortListener() {
            @Override
            public void onRead(String rawData) {


                String[] lines = rawData.split("\n");


                for (String line : lines) {

                    try {

                        line = line.replace("\r\n", "").trim();
                        if (line.startsWith("[") && line.endsWith("]")) {
                            line = line.substring(1, line.length() - 1);

                            String[] device = line.split(":");
                            switch (device[0]) {

                                case "sonarA0": {
                                    String cm = device[1]
                                            .replace("cm", "");
                                    int newReading = Integer.valueOf(cm);

                                    if (newReading != Devices.sonarA0.getDistanceToObject()) {
                                        Devices.sonarA0.setDistanceToObject(newReading);
                                        observer.onNext(SensorEvent.sonarDistanceMeasured);
                                    }
                                    break;
                                }
                                case ("servoA0"): {
                                    int newReading = Integer.valueOf(device[1]);
                                    if (Devices.servoA0.getCurrentPosition() != newReading) {
                                        Devices.servoA0.setCurrentPosition(newReading);
                                        observer.onNext(SensorEvent.servoMoved);
                                    }
                                    break;
                                }
                                case("servoB0") :
                                {
                                    int newReading = Integer.valueOf(device[1]);
                                    if (Devices.servoB0.getCurrentPosition() != newReading) {
                                        Devices.servoB0.setCurrentPosition(newReading);
                                        observer.onNext(SensorEvent.servoMoved);
                                    }

                                    break;
                                }





                            }


                        } else {
                            System.out.println("ignoring malformed line: " + line);
                        }


                    } catch (NumberFormatException ex) {
                        ex.printStackTrace();
                    }
                }
            }

            @Override
            public void onError(Throwable throwable) {

                throwable.printStackTrace();
            }
        });

        serialPortIO.start();

    }


    //BEN TODO - don't delete this - it has some logic in getting the alalog feedback from servos
//    SerialPortListener serialPortListener = new SerialPortListener() {
//        @Override
//        public void onRead(String rawData) {
//
//            String[] parts = rawData.split(",");
//            //   console.print(new Date().toString() +  " - serial data in - " + rawData);
//
//            for (String s : parts) {
//                s = s.replace("[", "").replace("]", "");
//                //     console.println(s);
//                String[] m = s.split(":");
//                if (m.length == 2) {
//
//                    if (m[0].equals("X")) {
//                        double d = Double.valueOf(m[1]) * 100;
//
//                        Integer v = (int) d;
//                        console.println("change in x " + v);
//                        if (! v.equals(lastX)) {
//
//                            lastX = v;
//
//                            int newPos = lastXPos + v;
//                            if (newPos < -90) {
//                                newPos = -90;
//                            }
//                            else if (newPos > 90) {
//                                newPos = 90;
//                            }
//
//                            lastXPos= newPos;
//                            console.println("moving on X " + lastXPos);
//                            servoController.angle(LF_Y, newPos).subscribe();
//                        }
//                    }
//
//                }
//            }
//
//
//        }
//
//        @Override
//        public void onError(Throwable throwable) {
//            console.print(throwable.getMessage());
//        }
//    };


}

File: src/com/colony/ArduinoSerialInterface.kt
package com.colony

import com.colony.model.Command
import com.colony.model.CommandType
import com.colony.serial.SerialPortIO
import com.colony.serial.SerialPortListener
import com.google.gson.GsonBuilder
import java.io.IOException
import java.lang.Exception
import java.util.*

/**
 * get serial data from arduino connected to usb port and post feedback to main program
 */
class ArduinoSerialInterface {
    private var serial: SerialPortIO? = null
    private val gson = GsonBuilder().create()
    private val random = Random()

    private var moved = false

    fun startListening() {
        serial = SerialPortIO(object : SerialPortListener {
            override fun onRead(rawData: String) {
//                println("read: $rawData")

                if (rawData.length > 0) {
                    val data = rawData.split('\n')
                    for (s in data) {
                        val cleanString = s.trim('\r')
                        when {
                            cleanString.isEmpty() -> {
                                return
                            }
                            cleanString.startsWith("DEBUG::") -> {
                               // println(cleanString)
                            }
                            else -> {
                                try {

                                    val command = gson.fromJson(cleanString, Command::class.java)


                                    val commandType: CommandType? = CommandType.lookup(command.id)

                                    when (commandType) {
                                        CommandType.SONAR -> {
                                           // println("Sonar Range: ${command.value}")
                                            if (!moved && command.value < 10) {
                                                val r = random.nextInt(160)
                                                println("obstacle detected moving to $r")
                                                moved = true
                                                sendCommand(Command(CommandType.SEEK.id, "", r))
                                            } else if (command.value > 10) {
                                                moved = false
                                            }
                                        }
                                        CommandType.PING -> {
                                            println("PONG $cleanString")
                                        }
                                        CommandType.SEEK -> TODO()
                                        CommandType.CALIBRATE -> TODO()
                                        null -> TODO()
                                    }


                                } catch (ex : Exception) {
                                    println("error processing input: $cleanString")
                                }
                            }
                        }
                    }
                }







            }

            override fun onError(throwable: Throwable) {
                throwable.printStackTrace()
            }
        })
        serial!!.start()
        try {
            Thread.sleep(2000)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
    }



    @Throws(IOException::class)
    fun sendCommand(command: Command) {
        val gson = GsonBuilder().create()
        val cmd = gson.toJson(command)
        val commandType = CommandType.lookup(command.id)
      //  println("sending $commandType command $cmd")
        serial!!.write(cmd)
        serial!!.write("\n")
        Thread.sleep(1000)

    }
}